zipline/utils/input_validation.py

```diff
@@ -30,6 +32,98 @@ def ensure_upper_case(func, argname, arg):
         )
 
 
+def ensure_dtype(func, argname, arg):
+    """
+    Argument preprocessor that converts the input into a numpy dtype.
+
+    Usage
+    -----
+    >>> import numpy as np
+    >>> from zipline.utils.preprocess import preprocess
+    >>> @preprocess(dtype=ensure_dtype)
+    ... def foo(dtype):
+    ...     return dtype
+    ...
+    >>> foo(float)
+    dtype('float64')
+    """
+    try:
+        return dtype(arg)
+    except TypeError:
+        raise TypeError(
+            "{func}() couldn't convert argument "
+            "{argname}={arg!r} to a numpy dtype.".format(
+                func=_qualified_name(func),
+                argname=argname,
+                arg=arg,
+            ),
+        )
+
+
+def expect_dtypes(*_pos, **named):
+    """
+    Preprocessing decorator that verifies inputs have expected numpy dtypes.
+
+    Usage
+    -----
+    >>> from numpy import dtype, arange
+    >>> @expect_dtypes(x=dtype(int))
+    ... def foo(x, y):
+    ...    return x, y
+    ...
+    >>> foo(arange(3), 'foo')
+    (array([0, 1, 2]), 'foo')
+    >>> foo(arange(3, dtype=float), 'foo')
+    Traceback (most recent call last):
+       ...
+    TypeError: foo() expected an argument with dtype 'int64' for argument 'x', but got dtype 'float64' instead.  # noqa
+    """
+    if _pos:
+        raise TypeError("expect_dtypes() only takes keyword arguments.")
+
+    for name, type_ in iteritems(named):
+        if not isinstance(type_, (dtype, tuple)):
+            raise TypeError(
+                "expect_dtypes() expected a numpy dtype or tuple of dtypes"
+                " for argument {name!r}, but got {dtype} instead.".format(
+                    name=name, dtype=dtype,
+                )
+            )
+    return preprocess(**valmap(_expect_dtype, named))
+
+
+def _expect_dtype(_dtype_or_dtype_tuple):
+    """
+    Factory for dtype-checking functions that work the @preprocess decorator.
+    """
+    # Slightly different messages for dtype and tuple of dtypes.
+    if isinstance(_dtype_or_dtype_tuple, tuple):
+        allowed_dtypes = _dtype_or_dtype_tuple
+    else:
+        allowed_dtypes = (_dtype_or_dtype_tuple,)
+    template = (
+        "%(funcname)s() expected a value with dtype {dtype_str} "
+        "for argument '%(argname)s', but got %(actual)r instead."
+    ).format(dtype_str=' or '.join(repr(d.name) for d in allowed_dtypes))
+
+    def check_dtype(value):
+        return getattr(value, 'dtype', None) not in allowed_dtypes
+
+    def display_bad_value(value):
+        # If the bad value has a dtype, but it's wrong, show the dtype name.
+        if hasattr(value, 'dtype'):
+            return value.dtype.name
```

Comment:
there are some cases where this will raise an attribute error, maybe check `hasattr(value, 'dtype') and hasattr(value.dtype, 'name')`